src/runtime/current_thread/runtime.rs

```diff
@@ -0,0 +1,111 @@
+use executor::current_thread::{self, CurrentThread};
+
+use tokio_reactor::{self, Reactor};
+use tokio_timer::timer::{self, Timer};
+use tokio_executor;
+
+use futures::Future;
+
+use std::io;
+
+/// Single-threaded runtime provides a way to start reactor
+/// and executor on the same thread.
+///
+/// The single-threaded runtime is not `Send` and cannot be
+/// safely moved to other threads.
+///
+/// # Examples
+///
+/// Creating a new `Runtime` with default configuration values.
+///
+/// ```
+/// use tokio::runtime::current_thread::Runtime;
+/// use tokio::prelude::*;
+///
+/// let mut runtime = Runtime::new().unwrap();
+///
+/// // Use the runtime...
+/// // runtime.block_on(f); // where f is a future
+/// ```
+#[derive(Debug)]
+pub struct Runtime {
+    reactor_handle: tokio_reactor::Handle,
+    timer_handle: timer::Handle,
+    executor: CurrentThread<Timer<Reactor>>,
+}
+
+impl Runtime {
+    /// Returns a new single-threaded runtime initialized with default
+    /// configuration values.
+    pub fn new() -> io::Result<Runtime> {
+        // We need a reactor to receive events about IO objects from kernel
+        let reactor = Reactor::new()?;
+        let reactor_handle = reactor.handle();
+
+        // Place a timer wheel on top of the reactor. If there are no timeouts to fire, it'll let the
+        // reactor pick up new some new external events.
+        let timer = Timer::new(reactor);
+        let timer_handle = timer.handle();
+
+        // And now put a single-threaded executor on top of the timer. When there are no futures ready
+        // to do something, it'll let the timer or the reactor to generate some new stimuli for the
+        // futures to continue in their life.
+        let executor = CurrentThread::new_with_park(timer);
+
+        let runtime = Runtime { reactor_handle, timer_handle, executor };
+        Ok(runtime)
+    }
+
+    /// Blocks and runs the given future.
+    ///
+    /// This is similar to running a runtime, but uses only the current thread.
+    pub fn block_on<F: Future<Item = (), Error = ()>>(&mut self, f: F)
+            -> Result<(), current_thread::BlockError<()>> {
```

Comment:
`BlockError` should probably be redefined in `runtime::current_thread`. This will allow us the ability to extend the error type when there are error conditions that only happen with the runtime and not the standalone executor.

Reply:
Let's do it next PR?..

Reply:
Ideally, master would always be able to be published to crates.io. So, if you want you can remove `block_on` entirely until a later PR. Either that, or we should fix the return type.

Reply:
Again, you already did a lot, so if you want I can do the rest.

Reply:
Yes please! :crab:

Reply:
I made `block_on` be the same signature as `block_on` from #328: https://github.com/tokio-rs/tokio/pull/328/files#diff-5dc86325a3b9c862b64f246354d78ffaR377